File: algorithms/javascript/sorting/selection.js


// selection sort

// this algorithm loop into the whole array, finds the smalled number and swap it with the first element of unsorted array
/*
let us say we have array [5,2,1,3,6,4]
and since we have not sorted anything yet the first element of unsorted array points on 0 index
and we find the smalled element which is `1` swap it with element at index 0 which is `5`
so the array looks like [1,2,5,3,6,4]
and since element at index 0 is sorted, the first element of the unsorted array is 1 which points to `2`
*/


// O(N^2)
function selectionSort(A){
    let smallesIndex = -1;

    for(let i=0; i< A.length; i++){
        smallesIndex = i

        // find the smallest number
        for(let j=i; j < A.length; j++)
            if(A[j] < A[smallesIndex]){
                smallesIndex = j
            }

        // swapp the found numbers
        [A[i],A[smallesIndex] ]= [A[smallesIndex],A[i]]
    }

    return A;
}


console.log(selectionSort([5,2,1,3,6,4]))
console.log(selectionSort([5,9,8,2,1,7,3,6,4]))
console.log(selectionSort([5291,20,1000,32,61,400]))

File: algorithms/cpp/sorting/selection.cpp
#include <iostream>
#include <vector>

using namespace std;

/**
 * @brief insertion sort takes it simple by going into the entire array and find the smalles value
 * when found we swap it with the the first element of array
 * and then since the first element is sorted we point on second element
 * we search with the rest of the array and if we find the element which is smaller than the pointed element we swap them unless
 * we do nothing we repeat the proccess until we point on the last element
 * 
 * @param array 
 * @return vector<int> 
 */
vector<int> selectionSort(vector<int> array){
    int size = array.size();
    int min = array.at(0);
    int minIndex = 0;
	bool toSwap = false;

    for(int i=0; i<size; i++){
		min = array[i];

        for(int j=i; j<size; j++){
            if(array[j] < min){
                minIndex = j;
				min = array.at(j);
				toSwap = true;
			}
        }


		if(toSwap)
	        swap(array[i],array[minIndex]);
		toSwap = false;

		
    }

	return array;
}

int main(){
	vector<int> array = {3,4,2,0,3,11,32};

	for(int a: selectionSort(array))
		cout<< a << "," ;


	cout << endl;

    return 0;
}